### PR TITLE
Add a 'wrap_to_coroutine' decorator to unify interfaces...

### PR DESCRIPTION
**Pull request by greschd**
…for functions and coroutines.

### CODE REVIEW

**Comment by mskoenz** [comment]
```python
@wrap_to_coroutine
async def async_snek():
     asyncio.sleep(1.) # I changed the 0 to a 1
     return '≻:======>···'
```

the sleep cmd would not work, since it is not awaited, right? To me it looks like an useless statement in the example...

**Comment by mskoenz** [comment]
Not a big fan of all these capital letters in the example, would prefere a main or use_case fct that can use regular var-names like one would use in the real world...

```python
if __name__ == '__main__':
    # would prefere here: demo_fct(), and put code below in that fct
    LOOP = asyncio.get_event_loop()
    FUT_SYNC = asyncio.ensure_future(sync_snek())
    FUT_ASYNC = asyncio.ensure_future(async_snek())
    LOOP.run_until_complete(asyncio.gather(FUT_SYNC, FUT_ASYNC))

    print('sync snek ', FUT_SYNC.result())
    print('async snek', FUT_ASYNC.result())
```

If it is only used in the doc, the `__main__` part could be dropped

**Comment by mskoenz** [comment]
make sure that the wrapped function still has doc-str and signature of original one. If you want, add a disclaimer to the doc str, but introspection-capabilities should not be broken.

I would suggest `functools.wraps`, not sure how nice it plays with `async`...

**Comment by mskoenz** [comment]
rest looks fine 👍

**Review by mskoenz — APPROVED** [review]

**Comment by greschd** [comment]
Example changes: makes sense :+1:

About the doc-string and signature: ``functools.wraps`` works, but afaik doesn't preserve the signature, only the docstring.

P.S.: If you comment directly in code, we can have separate threads of discussion about each issue.

**Comment by mskoenz** [comment]
`functools.wraps` is compatible with `inspect.signature`, i.e. if you use it, `inspect` will still find the original sig (afaik it just checks if there is a `__wrapped__` member and follows this)

File: examples/wrap_to_coroutine.py [hunk 1]
```diff
@@ -15,15 +15,22 @@ def sync_snek():
 
 @wrap_to_coroutine
 async def async_snek():
-    asyncio.sleep(0.)
+    await asyncio.sleep(0.)
```
Comment (mskoenz) [inline examples/wrap_to_coroutine.py]: I see

Thats how comments in code work... how about waiting longer than 0? I know its not necessary for your point, but this way, every programmer has the urge to remove this statement

**Comment by greschd** [comment]
Ah, I was checking ``inspect.getfullargspec``. Anyhow, I can just use ``decorator``, as it seems to also work with coroutines. I guess the parts of the code that ``decorator`` touches are the same for regular functions and coroutines.

**Comment by mskoenz** [comment]
I would prefer `functools.wraps`, since `decorator` is much more aggressive with magic

**Comment by mskoenz** [comment]
`inspect.getfullargspec` is depr since at least python3.5

**Comment by greschd** [comment]
``getfullargspec`` documentation:

> Changed in version 3.6: This method was previously documented as deprecated in favour of signature() in Python 3.5, but that decision has been reversed in order to restore a clearly supported standard interface for single-source Python 2/3 code migrating away from the legacy getargspec() API.

**Comment by mskoenz** [comment]
Alright, didn't know that yet... if `@wraps` does not work with `inspect.getfullargspec`, then `decorator.decorator` is fine

**Comment by greschd** [comment]
So, it's deprecated since at least 3.5, and 'un-deprecated' since at least 3.6 :rofl:

**Comment by greschd** [comment]
Alright, thanks for the review!

**Pull request merged**